arch/xtensa/core/xtensa-win0.S

```diff
@@ -0,0 +1,523 @@
+#include <xtensa/config/core-isa.h>
+#include <syscall_list.h>
+#include <zsr.h>
+#include <offsets.h>
+#include <syscall_list.h>
+
+/* Saves the "odd" thread context registers (thread state not captured
+ * by GPRs) to a context struct
+ */
+.macro ODD_REG_SAVE val_reg ctx_reg
+	rsr \val_reg, SAR
+	s32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_sar_OFFSET
+#if XCHAL_HAVE_LOOPS
+	rsr \val_reg, LBEG
+	s32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_lbeg_OFFSET
+	rsr \val_reg, LEND
+	s32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_lend_OFFSET
+	rsr \val_reg, LCOUNT
+	s32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_lcount_OFFSET
+#endif
+#if XCHAL_HAVE_S32C1I
+	rsr \val_reg, SCOMPARE1
+	s32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_scompare1_OFFSET
+#endif
+#if XCHAL_HAVE_THREADPTR
+	rur \val_reg, THREADPTR
+	s32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_threadptr_OFFSET
+#endif
+.endm
+
+/* Similarly, restores the "odd" registers from a thread context */
+.macro ODD_REG_RESTORE val_reg ctx_reg
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_sar_OFFSET
+	wsr \val_reg, SAR
+#if XCHAL_HAVE_LOOPS
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_lbeg_OFFSET
+	wsr \val_reg, LBEG
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_lend_OFFSET
+	wsr \val_reg, LEND
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_lcount_OFFSET
+	wsr \val_reg, LCOUNT
+#endif
+#if XCHAL_HAVE_S32C1I
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_scompare1_OFFSET
+	wsr \val_reg, SCOMPARE1
+#endif
+#if XCHAL_HAVE_THREADPTR
+	l32i \val_reg, \ctx_reg, __xtensa_win0_ctx_t_threadptr_OFFSET
+	wur \val_reg, THREADPTR
+#endif
+.endm
+
+/* Inc/decrement nested interrupt count on the CPU struct */
+.macro MOD_NEST_COUNT val cpu_reg scratch_reg
+	l32i \scratch_reg, \cpu_reg, ___cpu_t_nested_OFFSET
+	addi \scratch_reg, \scratch_reg, \val
+	s32i \scratch_reg, \cpu_reg, ___cpu_t_nested_OFFSET
+.endm
+
+/* On kernel entry, sets RING to zero (enabling kernel memory access)
+ * and clears EXCM (enabling further nested exceptions and interrupts)
+ */
+.macro PS_ENTRY_UNMASK ps_reg val_reg
+	rsr \ps_reg, PS
+	movi \val_reg, ~(PS_EXCM_MASK | PS_RING_MASK)
+	and \ps_reg, \ps_reg, \val_reg
+	wsr \ps_reg, PS
+.endm
+
+/* Vector handlers: the vector sections themselves are very small (~10
+ * instructions on some devices!), so they do minimal work.  Their job:
+ *
+ * + Place the interrupted a0 register into ZSR_A0SAVE
+ * + Place a C handler address corresponding to the interrupt in a0
+ * + Store the interrupted PC in ZSR_EPC
+ * + Store the interrupted PS register value in ZSR_EPS
+ * + Jump to the shared entry code
+ *
+ * Note that the User and Kernel Level 1 handlers must also check for
+ * two special entry paths (so they do the bulk of their work in a
+ * "l1_handler" section outside the vector region):
+ *
+ * + System calls have their own optimized handler
+ * + MMU TLB exceptions are handled directly, as they must not
+ *   re-enable interrupts at any point before returning to the trapping
```

Comment:
what you mean here ? syscall handler is expected to run with interruptions enabled
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/kernel/mem_protect/userspace/src/main.c#L934

Reply:
Maybe the typography is confusing?  Two items in a bulleted list:

* System calls have their own optimizer handler
* TLB exceptions don't unmask EXCM

Syscalls absolutely do unmask interrupts, obviously.